### PR TITLE
Block Entities article

### LINKED ISSUE #35: Developer Guides: Block Entities

### PR DESCRIPTION
**Pull request by natri0**
Describes how to:
- [x] register & create a block entity
- [x] 'bind' it to the block w/ `BlockEntityProvider` / `BlockWithEntity`
- [x] serialize/deserialize & use custom fields in the block entity
- [x] tick

(Maybe) To do:
- [ ] custom rendering w/ `BlockEntityRenderer`
- [ ] handled screens
- [ ] networking
- [ ] data components
- [ ] container blocks

Also, idk if the article/code should be copied to 1.20.4 if there's pretty much nothing changed as far as i can tell.

### CODE REVIEW

**Comment by its-miroma** [comment]
Thank you for your contribution!

> Also, idk if the article/code should be copied to 1.20.4 if there's pretty much nothing changed as far as i can tell.

We usually don't, but if there were no changes why not

**Comment by IMB11** [comment]
>  custom rendering w/ BlockEntityRenderer
 > handled screens
 > networking
 > data components
 > container blocks
 
 These can be subpages, no need to do them in this PR - this is a great start, i'll have a look at it now.

**Comment by IMB11** [comment]
Also, you don't need to backport changes to previous versions, the older versions are set in stone and are not really meant to be expanded upon.

File: develop/blocks/block-entities.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+title: Block Entities
+description: Learn how to create block entities for your custom blocks.
+authors:
+  - natri0
+---
+
+# Block Entities {#block-entities}
+
+Block entities are a way to store additional data for a block, that is not part of the block state: inventory contents, custom name and so on.
+Vanilla Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
```
Comment (IMB11) [inline develop/blocks/block-entities.md]:
```suggestion
Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
```

File: develop/blocks/block-entities.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+In this article, we will create a block that counts how many times it has been right-clicked.
```
Comment (IMB11) [inline develop/blocks/block-entities.md]:
```suggestion
As an example, we will create a block that counts how many times it has been right-clicked.
```

File: develop/blocks/block-entities.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+## Creating the Block Entity {#creating-the-block-entity}
+
+To make Minecraft "see" the new block entity, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it in the `ModBlockEntities` class, just like we did with the block.
```
Comment (IMB11) [inline develop/blocks/block-entities.md]:
```suggestion
To make Minecraft recognize and load the new block entities, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it a new `ModBlockEntities` class.
```

File: develop/blocks/block-entities.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+@[code transcludeWith=:::1](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/CounterBlockEntity.java)
+
+Registering a Block Entity yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:
```
Comment (IMB11) [inline develop/blocks/block-entities.md]:
```suggestion
Registering a `BlockEntity` yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:
```

File: develop/blocks/block-entities.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+The `BlockEntityProvider` interface also defines a method called `getTicker`, which can be used to run code every tick for each instance of the block. We can implement that by creating a static method that will be used as the `BlockEntityTicker`:
+
+The `getTicker` method should also check if the passed `BlockEntityType` is the same as the one we're using, and if it is, return the function that will be called every tick. Thankfully, there is a utility function that does the check in `BlockWithEntity`:
+
+@[code transcludeWith=:::3](@/reference/latest/src/main/java/com/example/docs/block/custom/CounterBlock.java)
+
+`CounterBlockEntity::tick` is a reference to the static method `tick` we should create in the `CounterBlockEntity` class. Structuring it like this is not required, but it's a good practice to keep the code clean and organized.
+
+Let's say we want to make it so that the counter can only be incremented once every 10 ticks (2 times a second). We can do this by adding a `ticksSinceLast` field to the `CounterBlockEntity` class, and increasing it every tick:
```
Comment (IMB11) [inline develop/blocks/block-entities.md:95]: Maybe mention something like "Don't forget to serialize and deserialize this new field!"

**Review by IMB11 — CHANGES_REQUESTED** [review]
Really good structure, just a few minor changes i'd suggest.

**Review by IMB11 — APPROVED** [review]

**Pull request merged**